miniapps/nurbs/nurbs_solenoidal.cpp

```diff
@@ -0,0 +1,398 @@
+   //    (u,p) and the linear forms (fform, gform).
+   MemoryType mt = device.GetMemoryType();
+   BlockVector x(block_offsets, mt), rhs(block_offsets, mt);
+   rhs = 0.0;
+
+   LinearForm *fform(new LinearForm);
+   fform->Update(R_space, rhs.GetBlock(0), 0);
+   fform->AddDomainIntegrator(new VectorFEDomainLFIntegrator(ucoeff));
+   fform->Assemble();
+   fform->SyncAliasMemory(rhs);
+
+   // 9. Assemble the finite element matrices for the Darcy operator
+   //
+   //                            D = [ M  B^T ]
+   //                                [ B   0  ]
+   //     where:
+   //
+   //     M = \int_\Omega k u_h \cdot v_h d\Omega   u_h, v_h \in R_h
+   //     B   = -\int_\Omega \div u_h q_h d\Omega   u_h \in R_h, q_h \in W_h
+   BilinearForm *mVarf(new BilinearForm(R_space));
+   MixedBilinearForm *bVarf(new MixedBilinearForm(R_space, W_space));
+
+   mVarf->AddDomainIntegrator(new VectorFEMassIntegrator(one));
+   mVarf->Assemble();
+   mVarf->Finalize();
+
+   bVarf->AddDomainIntegrator(new VectorFEDivergenceIntegrator);
+   bVarf->Assemble();
+   bVarf->Finalize();
+
+   SparseMatrix &M(mVarf->SpMat());
+   SparseMatrix &B(bVarf->SpMat());
+   B *= -1.;
+   TransposeOperator *Bt = new TransposeOperator(&B);
+
+   BlockOperator darcyOp(block_offsets);
+   darcyOp.SetBlock(0,0, &M);
+   if (div_free) { darcyOp.SetBlock(0,1, Bt); }
+   if (div_free) { darcyOp.SetBlock(1,0, &B); }
+
+   // 10. Construct the operators for preconditioner
+   //
+   //                 P = [ diag(M)         0         ]
+   //                     [  0       B diag(M)^-1 B^T ]
+   //
+   //     Here we use Symmetric Gauss-Seidel to approximate the inverse of the
+   //     pressure Schur Complement
+   Vector Md(mVarf->Height());
+   M.GetDiag(Md);
+   Md.HostReadWrite();
+   SparseMatrix *MinvBt = Transpose(B);
+   for (int i = 0; i < Md.Size(); i++)
+   {
+      MinvBt->ScaleRow(i, 1./Md(i));
+   }
+   SparseMatrix *S = Mult(B, *MinvBt);
+   Solver *invS;
+#ifndef MFEM_USE_SUITESPARSE
+   invS = new GSSmoother(*S);
+#else
+   invS = new UMFPackSolver(*S);
+#endif
+   invS->iterative_mode = false;
+
+   Solver *invM = new GSSmoother(M);
+   invM->iterative_mode = false;
+
+   BlockDiagonalPreconditioner darcyPrec(block_offsets);
+   darcyPrec.SetDiagonalBlock(0, invM);
+   darcyPrec.SetDiagonalBlock(1, invS);
+
+   // 11. Solve the linear system with MINRES.
+   //     Check the norm of the unpreconditioned residual.
+   int maxIter(10000);
+   double rtol(1.e-16);
+   double atol(1.e-16);
+
+   chrono.Clear();
+   chrono.Start();
+   MINRESSolver solver;
+   solver.SetAbsTol(atol);
+   solver.SetRelTol(rtol);
+   solver.SetMaxIter(maxIter);
+   solver.SetOperator(darcyOp);
+   solver.SetPreconditioner(darcyPrec);
+   solver.SetPrintLevel(2);
+   x = 0.0;
+   solver.Mult(rhs, x);
+   if (device.IsEnabled()) { x.HostRead(); }
+   chrono.Stop();
+
+   if (solver.GetConverged())
+   {
+      std::cout << "MINRES converged in " << solver.GetNumIterations()
```

Comment:
There is usage of `mfem::out` and `std::cout`. I'm not sure it matters, but it is inconsistent. The examples usually just use `cout`.

Reply:
I think you are correct. In the library it is a must. In executables `std::cout` is still allowed, apparently.
I changed it any way- as suggested- for consistency.